src/basic/fileio.c

```diff
@@ -310,6 +310,100 @@ int verify_file(const char *fn, const char *blob, bool accept_extra_nl) {
         return 1;
 }
 
+int read_full_virtual_file(const char *filename, char **ret_contents, size_t *ret_size) {
+        _cleanup_free_ char *buf = NULL;
+        _cleanup_close_ int fd = -1;
+        struct stat st;
+        ssize_t n, size;
+        char *p;
+
+        assert(ret_contents);
+
+        /* Virtual filesystems such as sysfs or procfs use kernfs, and kernfs can work
+         * with two sorts of virtual files. One sort uses "seq_file", and the results of
+         * the first read are buffered for the second read. The other sort uses "raw"
+         * reads which always go direct to the device. In the latter case, the content of
+         * the virtual file must be retrieved with a single read otherwise a second read
+         * might get the new value instead of finding EOF immediately. That's the reason
+         * why the usage of fread(3) is prohibited in this case as it always performs a
+         * second call to read(2) looking for EOF. See issue 13585. */
+
+        fd = open(filename, O_RDONLY|O_CLOEXEC);
+        if (fd < 0)
+                return -errno;
+
+        /* Start size for files in /proc which usually report a file size of 0. */
+        size = LINE_MAX;
+
+        for (;;) {
```

Comment:
Shouldn't there be maximum number of tries here?

Reply:
The only way it would loop forever would be if the kernel constantly reports an incorrect file size (more data read than the reported file size)... I don't think it could happen in practice but maybe we could add an extra safety check.

Reply:
You're right, it would have to grow. But we use stat, so it could grow just by one byte every time. So I think it's better to have some limit here.

Reply:
Ok I'll add it.